### PR TITLE
subscription period

### PR DESCRIPTION
**Pull request by aboedo**
Adds subscripion period to the POST receipt request. 

Changes summary: 

- moved the test files into groups that mimic the ones for the main repo, for organization purposes. 
- added the fields from https://github.com/RevenueCat/khepri/pull/1539 to the post receipt request:
    - `normal_duration`
    - `intro_duration`
    - `trial_duration`
- The number of parameters for the post receipt endpoint was getting pretty unwieldy, and would change any time we read a new field from SKProducts. I extracted a few of those fields into a new class, `RCProductInfo`. This class also knows how to convert itself into a backend-compatible dictionary, and how to generate a cache key. 
- moved the logic that reads individual fields from `SKProducts` and does whatever transformations are needed for the backend endpoint into `RCProductInfoExtractor`. This also abstracts the logic of which fields can be read in which iOS versions. 
- Some minor refactors here and there. But the main changes are the ones described above.

### CODE REVIEW

File: Examples/SwiftExample/Podfile.lock [hunk 1]
```diff
@@ -9,8 +9,8 @@ EXTERNAL SOURCES:
     :path: "../../"
 
 SPEC CHECKSUMS:
-  Purchases: 0ed262862309e8aa3348b525755c8e84f53b1924
+  Purchases: 6bf68ed9926fefa775649bec025878c4de9175b5
 
 PODFILE CHECKSUM: b6ad00ca1e6fc400e0898ad2ac4a293958616d2e
 
-COCOAPODS: 1.8.4
+COCOAPODS: 1.9.1
```
Comment (aboedo) [inline Examples/SwiftExample/Podfile.lock:16]: just the result of upgrading cocoapods and running pod install.

File: Purchases.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -13,20 +13,16 @@
...
 		2D50332B2406EA61009CAE61 /* RCSubscriberAttributesManager.m in Sources */ = {isa = PBXBuildFile; fileRef = 2D5033282406EA61009CAE61 /* RCSubscriberAttributesManager.m */; };
 		2D8DB34B24072AAE00BE3D31 /* SubscriberAttributeTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2D8DB34A24072AAE00BE3D31 /* SubscriberAttributeTests.swift */; };
-		2D8DB3502407333100BE3D31 /* NSError+RCExtensionsTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 37E3550459807FF5636B2667 /* NSError+RCExtensionsTests.swift */; };
 		2DD448FF24088473002F5694 /* RCPurchases+SubscriberAttributes.h in Headers */ = {isa = PBXBuildFile; fileRef = 2DD448FD24088473002F5694 /* RCPurchases+SubscriberAttributes.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		2DD4490024088473002F5694 /* RCPurchases+SubscriberAttributes.m in Sources */ = {isa = PBXBuildFile; fileRef = 2DD448FE24088473002F5694 /* RCPurchases+SubscriberAttributes.m */; };
+		2DEB9767247DB46900A92099 /* RCISOPeriodFormatter.h in Headers */ = {isa = PBXBuildFile; fileRef = 2DEB9766247DB46900A92099 /* RCISOPeriodFormatter.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Comment (aboedo) [inline Purchases.xcodeproj/project.pbxproj:18]: two main things happening in this file: 
- moved test files intro groups, essentially a follow-up of https://github.com/RevenueCat/purchases-ios/pull/242
- added new files: `RCISOPeriodFormatter`, `RCProductInfo`, `RCProductInfoExtractor`

File: Purchases/Networking/RCBackend.h [hunk 3]
```diff
@@ -11,25 +11,16 @@
...
 NS_ASSUME_NONNULL_BEGIN
 
 @class RCPurchaserInfo, RCHTTPClient, RCIntroEligibility, RCPromotionalOffer, RCSubscriberAttribute, RCSystemInfo;
 
-typedef NS_ENUM(NSInteger, RCPaymentMode) {
-    RCPaymentModeNone = -1,
-    RCPaymentModePayAsYouGo = 0,
-    RCPaymentModePayUpFront = 1,
-    RCPaymentModeFreeTrial = 2
-};
```
Comment (aboedo) [inline Purchases/Networking/RCBackend.h:24]: moved to `RCProductInfo`

File: Purchases/Networking/RCBackend.h [hunk 4]
```diff
@@ -54,13 +45,7 @@ typedef void(^RCOfferSigningResponseHandler)(NSString * _Nullable signature,
...
                   appUserID:(NSString *)appUserID
                   isRestore:(BOOL)isRestore
-          productIdentifier:(nullable NSString *)productIdentifier
-                      price:(nullable NSDecimalNumber *)price
-                paymentMode:(RCPaymentMode)paymentMode
-          introductoryPrice:(nullable NSDecimalNumber *)introductoryPrice
-               currencyCode:(nullable NSString *)currencyCode
-          subscriptionGroup:(nullable NSString *)subscriptionGroup
-                  discounts:(nullable NSArray<RCPromotionalOffer *> *)discounts
+                productInfo:(nullable RCProductInfo *)productInfo
```
Comment (aboedo) [inline Purchases/Networking/RCBackend.h:48]: abstracted parameters into object

File: Purchases/Networking/RCBackend.h [hunk 5]
```diff
@@ -11,25 +11,16 @@
...
-    RCPaymentModePayUpFront = 1,
-    RCPaymentModeFreeTrial = 2
-};
-
 extern NSErrorUserInfoKey const RCSuccessfullySyncedKey;
 extern NSString * const RCAttributeErrorsKey;
 extern NSString * const RCAttributeErrorsResponseKey;
 
-API_AVAILABLE(ios(11.2), macos(10.13.2))
-RCPaymentMode RCPaymentModeFromSKProductDiscountPaymentMode(SKProductDiscountPaymentMode paymentMode);
```
Comment (aboedo) [inline Purchases/Networking/RCBackend.h:31]: moved to `RCProductInfo`.

I'm not a huge fan of the format, I was going to turn it into a static method instead but changing it didn't seem to add much value anyway

File: Purchases/Public/RCPurchases.h [hunk 6]
```diff
@@ -270,7 +270,7 @@ NS_SWIFT_NAME(purchaserInfo(_:));
  @param completion An @escaping callback that is called with the loaded products. If the fetch fails for any reason it will return an empty array.
  */
 - (void)productsWithIdentifiers:(NSArray<NSString *> *)productIdentifiers
-                     completionBlock:(RCReceiveProductsBlock)completion
+                completionBlock:(RCReceiveProductsBlock)completion
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.h:273]: just indentation

File: Purchases/Public/RCPurchases.m [hunk 7]
```diff
@@ -336,8 +339,8 @@ - (void)setDelegate:(id<RCPurchasesDelegate>)delegate
 #pragma mark Attribution
 
 - (void)postAttributionData:(NSDictionary *)data
-               fromNetwork:(RCAttributionNetwork)network
-          forNetworkUserId:(nullable NSString *)networkUserId
+                fromNetwork:(RCAttributionNetwork)network
+           forNetworkUserId:(nullable NSString *)networkUserId
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:343]: indentation was off

File: Purchases/Public/RCPurchases.m [hunk 8]
```diff
@@ -624,36 +627,40 @@ - (void)restoreTransactionsWithCompletionBlock:(nullable RCReceivePurchaserInfoB
...
-                    subscriptionGroup:nil
-                            discounts:nil
+                          productInfo:nil
           presentedOfferingIdentifier:nil
                          observerMode:!self.finishTransactions
                  subscriberAttributes:subscriberAttributes
                            completion:^(RCPurchaserInfo *_Nullable info, NSError *_Nullable error) {
-                               [self dispatch:^{
-                                   if (error) {
-                                       [self markAttributesAsSyncedIfNeeded:subscriberAttributes
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:640]: some of the changes in this file are moving these completion blocks into functions for readability. the methods were pretty big otherwise, and it was hard to follow the flow

File: Purchases/Purchasing/RCProductInfo.m [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+// Copyright (c) 2020 Purchases. All rights reserved.
+//
+
+#import "RCProductInfo.h"
+#import "RCPromotionalOffer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+
+@interface RCProductInfo ()
```
Comment (aboedo) [inline Purchases/Purchasing/RCProductInfo.m:12]: most of this file is extracted from others and cleaned up a bunch

File: Purchases/Purchasing/RCProductInfo.m [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        durations[@"normal_duration"] = self.normalDuration;
+    }
+    if (self.introDurationType == RCIntroDurationTypeIntroPrice) {
+        durations[@"intro_duration"] = self.introDuration;
+    }
+    if (self.introDurationType == RCIntroDurationTypeFreeTrial) {
+        durations[@"trial_duration"] = self.introDuration;
+    }
+
+    return durations;
```
Comment (aboedo) [inline Purchases/Purchasing/RCProductInfo.m:136]: this is the only big new thing here

File: Purchases/Purchasing/RCProductInfoExtractor.m [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+- (instancetype)init {
+    if (self = [super init]) {
+        if (@available(iOS 11.2, macOS 10.13.2, tvOS 11.2, *)) {
+            self.formatter = [[RCISOPeriodFormatter alloc] init];
+        }
+    }
+    return self;
+}
+
+- (RCProductInfo *)extractInfoFromProduct:(SKProduct *)product {
```
Comment (aboedo) [inline Purchases/Purchasing/RCProductInfoExtractor.m:36]: most of this file was extracted from RCPurchases and cleaned up so that it's easier to follow

File: Purchases/Purchasing/RCProductInfoExtractor.m [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    NSDecimalNumber *introPrice = nil;
+    if (@available(iOS 11.2, macOS 10.13.2, tvOS 11.2, *)) {
+        if (product.introductoryPrice) {
+            introPrice = product.introductoryPrice.price;
+        }
+    }
+    return introPrice;
+}
+
+- (NSString *)extractNormalDurationForProduct:(SKProduct *)product {
```
Comment (aboedo) [inline Purchases/Purchasing/RCProductInfoExtractor.m:122]: this and the method after it are the real new things here

File: PurchasesTests/Mocks/MockSKProduct.swift [hunk 13]
```diff
@@ -4,16 +4,54 @@
 //
 
 class MockSKProduct: SKProduct {
```
Comment (aboedo) [inline PurchasesTests/Mocks/MockSKProduct.swift:6]: we had two separate mocks for SKProduct, `MockSKProduct` and `MockProduct`. 
I unified them, updated tests and added a whole bunch of fields / mock stuff

File: PurchasesTests/Purchasing/ProductInfoExtractorTests.swift [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    func testExtractInfoFromProductExtractsCurrencyCode() {
+        let product = MockSKProduct(mockProductIdentifier: "cool_product")
+        product.mockPriceLocale = Locale(identifier: "es_UY")
+        let productInfoExtractor = RCProductInfoExtractor()
+
+        var receivedProductInfo = productInfoExtractor.extractInfo(from: product)
+
+        expect(receivedProductInfo.currencyCode) == "UYU"
```
Comment (aboedo) [inline PurchasesTests/Purchasing/ProductInfoExtractorTests.swift:34]: Uruguayan Peso 💵 🇺🇾

File: PurchasesTests/Purchasing/ProductInfoExtractorTests.swift [hunk 15]
```diff
@@ -0,0 +1,194 @@
+import XCTest
+import Nimble
+
+import Purchases
+
+class ProductInfoExtractorTests: XCTestCase {
+
+    func testExtractInfoFromProductExtractsProductIdentifier() {
+        let product = MockSKProduct(mockProductIdentifier: "cool_product")
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoExtractorTests.swift:9]: I would extract `"cool_product"` to a `expected`,  so that you can assert against it in line 14

File: PurchasesTests/Purchasing/ProductInfoExtractorTests.swift [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+        let productInfoExtractor = RCProductInfoExtractor()
+
+        let receivedProductInfo = productInfoExtractor.extractInfo(from: product)
+
+        expect(receivedProductInfo.productIdentifier) == "cool_product"
+    }
+
+    func testExtractInfoFromProductExtractsPrice() {
+        let product = MockSKProduct(mockProductIdentifier: "cool_product")
+        product.mockPrice = NSDecimalNumber(decimal: 10.99)
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoExtractorTests.swift:19]: same here, I would have `let expected = 10.99`. This could apply to all the tests in this file

File: PurchasesTests/Purchasing/ProductInfoTests.swift [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+        let productInfo = RCProductInfo(productIdentifier: product_identifier,
+                                        paymentMode: .none,
+                                        currencyCode: "UYU",
+                                        price: 9.99,
+                                        normalDuration: nil,
+                                        introDuration: nil,
+                                        introDurationType: .none,
+                                        introPrice: nil,
+                                        subscriptionGroup: nil,
+                                        discounts: nil)
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoTests.swift:18]: could you make a function that builds an empty `RCProductInfo` and then modify each field that you're testing? Something like:

```
let productInfo = emptyProductInfo()
productInfo.productIdentifier = "cool_product"
```

File: PurchasesTests/Purchasing/ProductInfoTests.swift [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+        var productInfo = RCProductInfo(productIdentifier: "cool_product",
+                                        paymentMode: paymentMode,
+                                        currencyCode: "UYU",
+                                        price: 9.99,
+                                        normalDuration: nil,
+                                        introDuration: nil,
+                                        introDurationType: .none,
+                                        introPrice: nil,
+                                        subscriptionGroup: nil,
+                                        discounts: nil)
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoTests.swift:33]: In this test you could create the empty product info I mentioned above and then just modify the paymentMode to cover all cases

File: PurchasesTests/Purchasing/ProductInfoTests.swift [hunk 19]
```diff
@@ -0,0 +1,273 @@
+import XCTest
+import Nimble
+
+import Purchases
+
+class ProductInfoTests: XCTestCase {
+    func testAsDictionaryConvertsProductIdentifierCorrectly() {
+        let product_identifier = "cool_product"
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoTests.swift:8]:
```suggestion
        let productIdentifier = "cool_product"
```

File: PurchasesTests/Purchasing/ProductInfoTests.swift [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+                                        currencyCode: "UYU",
+                                        price: 9.99,
+                                        normalDuration: nil,
+                                        introDuration: introDuration,
+                                        introDurationType: .none,
+                                        introPrice: nil,
+                                        subscriptionGroup: nil,
+                                        discounts: nil)
+        expect(productInfo.asDictionary()["trial_duration"]).to(beNil())
+        expect(productInfo.asDictionary()["trial_duration"]).to(beNil())
```
Comment (vegaro) [inline PurchasesTests/Purchasing/ProductInfoTests.swift:169]:
```suggestion
        expect(productInfo.asDictionary()["intro_duration"]).to(beNil())
```

File: Purchases/Networking/RCBackend.m [hunk 21]
```diff
@@ -196,83 +175,39 @@ - (void)    postReceiptData:(NSData *)data
...
     }
 
     [self.httpClient performRequest:@"POST"
                                path:@"/receipts"
                                body:body
                             headers:self.headers
                   completionHandler:^(NSInteger status, NSDictionary *response, NSError *error) {
-                      for (RCBackendPurchaserInfoResponseHandler callback in [self getCallbacksAndClearForKey:cacheKey]) {
+                      for (RCBackendPurchaserInfoResponseHandler
+                          callback in [self getCallbacksAndClearForKey:cacheKey]) {
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/Networking/RCBackend.m:210]: AppCode's auto-formatter did it because the line crossed the 120-char limit. I'll improve it. 
It is normal, though, just not pretty.

File: PurchasesTests/Networking/BackendTests.swift [hunk 22]
```diff
@@ -1082,35 +1014,32 @@ class BackendTests: XCTestCase {
...
+        let discount = RCPromotionalOffer()
         discount.offerIdentifier = "offerid"
-        discount.paymentMode = RCPaymentMode.payAsYouGo
+        discount.paymentMode = .payAsYouGo
         discount.price = 12
-
+        let productInfo = RCProductInfo(productIdentifier: "product_id",
+                                        paymentMode: .none,
+                                        currencyCode: "UYU",
+                                        price: 15.99,
```
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesTests/Networking/BackendTests.swift:1032]: yeah, I did think about that when writing. the tests are definitely more noisy this way. 
The reason I went this way was: 
- I wanted _only_ arguments that were truly nullable (i.e.: you might not be able to extract them from an SKProduct) to be nullable. 
- I didn't want properties to be editable unless they needed to (this part could be done by doing a protected header, though)
- Even though you can initialize RCProductInfo with no arguments (because it inherits from NSObject), that can lead to some pretty bad gotchas with swift: 

When you initialize with no arguments, everything is nil. However, swift will assume that the object can't be nil unless nullable is specified (or no nullability is specified in the entire file). 

For example, if I do: 
```swift
let productInfo = RCProductInfo()
let price = productInfo.price
```
That will crash, because price is nil, but swift didn't expect that. I won't be able to even check for it, since `if price != nil` will raise a compiler warning saying that price can't be nil. 

To get around this, I usually disable the no-params constructor, I forgot about it this time. 
To disable it, you do `NS_UNAVAILABLE` ([example here](https://github.com/RevenueCat/purchases-ios/blob/develop/Purchases/Misc/RCSystemInfo.h#L16) which will raise a compiler warning when using that constructor. 

That way you have control over the properties to make sure that they're never called from swift without having initialized them properly. 

Thinking about possible fixes for this, I think the best workaround is to have a protected header that allows for these properties to be modified, and then a single function that always creates a valid ProductInfo object, so that I can do: 
```swift
let productInfo = createProductInfo()
productInfo.whateverProperty = "the value I want to test"
```

I'll do that refactor

File: PurchasesTests/Purchasing/ProductInfoTests.swift (hunk 17 above)
Comment (aboedo) [reply PurchasesTests/Purchasing/ProductInfoTests.swift:18]: related to https://github.com/RevenueCat/purchases-ios/pull/256#discussion_r436099550, 
I'll do the refactor, but it can't really contain an empty product info because of the reasons mentioned there

File: PurchasesTests/Purchasing/ProductInfoTests.swift (hunk 19 above)
Comment (aboedo) [reply PurchasesTests/Purchasing/ProductInfoTests.swift:8]: oops

File: PurchasesTests/Purchasing/ProductInfoTests.swift (hunk 20 above)
Comment (aboedo) [reply PurchasesTests/Purchasing/ProductInfoTests.swift:169]: I was clearly tired when I got to this file 😂

File: PurchasesTests/Purchasing/ProductInfoTests.swift (hunk 17 above)
Comment (vegaro) [reply PurchasesTests/Purchasing/ProductInfoTests.swift:18]: Yeah, I was actually referring to create a function that builds:

```RCProductInfo(productIdentifier: "cool_product",
                                        paymentMode: .none,
                                        currencyCode: "UYU",
                                        price: price,
                                        normalDuration: nil,
                                        introDuration: nil,
                                        introDurationType: .none,
                                        introPrice: nil,
                                        subscriptionGroup: nil,
                                        discounts: nil)
```

And then modify each property. It's not an `emptyProductInfo`, it's more a `validProductInfo`, as you mentioned 👍

File: PurchasesTests/Networking/BackendTests.swift (hunk 22 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesTests/Networking/BackendTests.swift]: done!

File: PurchasesTests/Purchasing/ProductInfoTests.swift (hunk 17 above)
Comment (aboedo) [reply PurchasesTests/Purchasing/ProductInfoTests.swift]: done!
